:sparkles: show non-compliant code changes for CI-Tests, Code-Review and SAST checks in --show-details mode

boahc077 opened the pull request:
#### What kind of change does this PR introduce?
Fix for https://github.com/ossf/scorecard/issues/2798
This change shows non-compliant code changes for CI-Tests, Code-Review and SAST checks when scorecard is run in `--show-details` mode.

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?
These checks do not show non-compliant code changes in `Details` field when scorecard is run in `--show-details` mode.

#### What is the new behavior (if this is a feature change)?**
These shows show non-compliant Pull Requests and/or Commit SHAs in `Details` field when scorecard is run in `--show-details` mode.

- [ ] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes
Fixes #2798

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?
Yes

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release,
include the string "ACTION REQUIRED".

For more information on release notes see: https://git.k8s.io/release/cmd/release-notes/README.md
-->

```release-note
show non-compliant code changes for CI-Tests, Code-Review and SAST checks in --show-details mode
```

Signed-off-by: Ashish Kurmi <akurmi@stepsecurity.io>

laurentsimon commented on checks/evaluation/ci_tests.go:
```diff
@@ -61,8 +61,8 @@ func CITests(name string, c *checker.CITestData, dl checker.DetailLogger) checke
 			// use only one commit (branch HEAD) to refer to all commits in a PR
-			dl.Debug(&checker.LogMessage{
-				Text: fmt.Sprintf("merged PR without CI test at HEAD: %s", r.HeadSHA),
+			dl.Warn(&checker.LogMessage{
```
is the reason for changing this because we want to have it available in the BQTable? If so, could we instead run the cron job with --verbosity debug?

@spencerschrock @azeemsgoogle wdut?

NOTE: we have to be careful how this interacts with SARIF results (for which we don't have good e2e tests atm). Every new `Warn` in the details may create a new alerts in the GitHub scanning dashboard https://github.com/ossf/scorecard/blob/main/pkg/sarif.go#L309. In the case of the code above, it won't happen because there's no "file" associated with the warning.

raghavkaul replied:
I believe you want `Info` here - `Info` displays with `--show-details` but probably doesn't interfere with SARIF/Cron.

laurentsimon replied:
Info is for positive things. Here we're reporting something that's negative, iiuc

laurentsimon replied:
Thanks!

laurentsimon commented on checks/evaluation/code_review.go:
```diff
@@ -54,15 +57,37 @@ func CodeReview(name string, dl checker.DetailLogger, r *checker.CodeReviewData)
 	}
 
+	// Let's include non-compliant revision IDs in details
+	if len(botReviewActivityRevisionIDs) > 0 {
```
same comment as above. Warn or Debug?

In the past, we've used Warn for things that can be remediated, and debug for things that provide additional information.

boahc077 replied:
Do Debug messages show up in CLI? In my limited experiment, it didn't show up so I decided to use Warn to explain why the user is getting a low score in this category.

spencerschrock replied:
you would need to run with both `--show-details` and `--verbosity debug`

spencerschrock replied:
Spoke with Raghav and the code review changes are fine here, we will work around them in the other PR.

I still have a question around collecting the reviewed bot PR

boahc077 replied:
Thanks @spencerschrock  for flatting the reviewed bot issue, I have removed that part.

raghavkaul commented:
Wrt. Code Review logging changes, no need to revert them - I'll rebase on top of this.

boahc077 commented on checks/sast.go:
```diff
@@ -124,15 +130,16 @@ func SAST(c *checker.CheckRequest) checker.CheckResult {
 
 	totalMerged := 0
 	totalTested := 0
+	nonCompliantPRs := []int{}
```
oh, didn't know that we can have duplicates here. I have updated code to use a map. I also verified this change with the command you have shared to make sure that it doesn't show duplicate values.

boahc077 commented on checks/sast.go:
```diff
@@ -116,6 +119,9 @@ func SAST(c *checker.CheckRequest) checker.CheckResult {
 			return checker.CreateMaxScoreResult(CheckSAST, "SAST tool is run on all commits")
 		}
 
+		c.Dlogger.Warn(&checker.LogMessage{
```
I have fixed this issue in my latest commit.

boahc077 commented:
> Overall looks good. Thanks for helping improve the results!
> 
> I've left 1-2 comments

@spencerschrock thanks Spencer for reviewing the PR and providing your feedback, I have addressed all of your comments in my latest PR.
